src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/ReadOnlySpanND.cs

```diff
@@ -0,0 +1,468 @@
+    /// </summary>
+    [DebuggerTypeProxy(typeof(SpanNDDebugView<>))]
+    [DebuggerDisplay("{ToString(),raw}")]
+    public readonly ref struct ReadOnlySpanND<T>
+    {
+        /// <summary>A byref or a native ptr.</summary>
+        internal readonly ref T _reference;
+        /// <summary>The number of elements this SpanND contains.</summary>
+        internal readonly nint _linearLength;
+        /// <summary>The lengths of each dimension.</summary>
+        internal readonly ReadOnlySpan<nint> _lengths;
+        /// <summary>The strides representing the memory offsets for each dimension.</summary>
+        private readonly ReadOnlySpan<nint> _strides;
+        /// <summary>If the backing memory is permanently pinned (so not just using a fixed statement).</summary>
+        private readonly bool _isPinned = false;
+
+        /// <summary>
+        /// Creates a new read-only span over the entirety of the target array.
+        /// </summary>
+        /// <param name="array">The target array.</param>
+        /// <param name="lengths">The <see cref="System.ReadOnlySpan{T}"/> with the lengths of each dimension.</param>
+        /// <remarks>Returns default when <paramref name="array"/> is null.</remarks>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ReadOnlySpanND(T[]? array, ReadOnlySpan<nint> lengths)
+        {
+            if (array == null)
+            {
+                this = default;
+                return; // returns default
+            }
+
+            _linearLength = SpanNDHelpers.CalculateTotalLength(lengths);
+            if (_linearLength != array.Length)
+                ThrowHelper.ThrowArgument_LengthsMustEqualArrayLength();
+
+            _reference = ref MemoryMarshal.GetArrayDataReference(array);
+            _linearLength = array.Length;
+            _lengths = lengths.ToArray().AsSpan();
+            _strides = SpanNDHelpers.CalculateStrides(Rank, lengths);
+        }
+
+        /// <summary>
+        /// Creates a new read-only span over the portion of the target array beginning
+        /// at 'start' index and ending at 'end' index (exclusive).
+        /// </summary>
+        /// <param name="array">The target array.</param>
+        /// <param name="start">The index at which to begin the read-only span.</param>
+        /// <param name="lengths">The <see cref="System.ReadOnlySpan{T}"/> with the lengths of each dimension.</param>
+        /// <remarks>Returns default when <paramref name="array"/> is null.</remarks>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="start"/> or end index is not in the range (&lt;0 or &gt;Length).
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ReadOnlySpanND(T[]? array, nint start, ReadOnlySpan<nint> lengths)
+        {
+            _linearLength = SpanNDHelpers.CalculateTotalLength(lengths);
```

Comment:
Same comment about the potential for torn state.

Reply:
Fixed.